### PR TITLE
Irepela/homework 10

### PR DESCRIPTION
**Pull request by winner6**

### CODE REVIEW

File: irepela/homework_10/generators.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    index = 0
+    while True:
+        if index >= len(iterable):
+            index = 0
+
+        try:
+            yield iterable[index]
+            index += 1
+        except IndexError:
+            break
```
Comment (kudinov-fedor) [inline irepela/homework_10/generators.py:26]: why would you expect  Index error if you check     index >= len(iterable) and reset cursor

File: irepela/homework_10/generators.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    while True:
+        if is_ascending and index == len(iterable) - 1:
+            is_ascending = False
+            index = len(iterable) - 1
+        if not is_ascending and index == 0:
+            is_ascending = True
+
+        try:
+            yield iterable[index]
+            index = (index + 1) if is_ascending else (index - 1)
```
Comment (kudinov-fedor) [inline irepela/homework_10/generators.py:41]: increment = 1 if ascending else -1
index += increment

File: irepela/homework_10/generators.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            index += 1
+        except IndexError:
+            break
+
+
+def ping_pong_generator(iterable):
+    index = 0
+    is_ascending = True
+    while True:
+        if is_ascending and index == len(iterable) - 1:
```
Comment (kudinov-fedor) [inline irepela/homework_10/generators.py:33]: I think you do not need to check if it is ascending or not,

if first index -> change to ascending
if last index -> change to descending

File: irepela/homework_10/generators.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            is_ascending = False
+            index = len(iterable) - 1
+        if not is_ascending and index == 0:
+            is_ascending = True
+
+        try:
+            yield iterable[index]
+            index = (index + 1) if is_ascending else (index - 1)
+        except IndexError:
+            break
```
Comment (kudinov-fedor) [inline irepela/homework_10/generators.py:43]: why should you expect index error if you controll cursor

File: irepela/homework_10/iterators.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+
+class CycleIterator(Iterator):
+
+    def __init__(self, iterable):
+        self.iterable = iterable
+        self.index = 0
+
+    def __next__(self):
+        while True:
```
Comment (kudinov-fedor) [inline irepela/homework_10/iterators.py:39]: you do not need while here

you return in try
you break in except

you will never have 2 turn

File: irepela/homework_10/iterators.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+class PingPongIterator(Iterator):
+
+    def __init__(self, iterable):
+        self.iterable = iterable
+        self.is_ascending = True
+        self.index = 0
+
+    def __next__(self):
+        while True:
```
Comment (kudinov-fedor) [inline irepela/homework_10/iterators.py:58]: you do not need loop here, it is always 1 pass - give 1 element and thats it

File: tests/irepela/homework_10/test_context_manager_timer.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+    with Timer() as timer:
+        time.sleep(3)
+
+    assert timer.total >= 3
+
+    with timer:
+        time.sleep(2)
+
+    assert timer.total >= 2
```
Comment (kudinov-fedor) [inline tests/irepela/homework_10/test_context_manager_timer.py:16]: in your case it is >= 5

File: tests/irepela/homework_10/test_context_manager_try_except.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+def test_try_except():
+    with Catcher():
+        5 / 0
+
+    with Catcher(ZeroDivisionError):
+        5 / 0
+
+    with Catcher(TypeError, ZeroDivisionError):
+        5 / 0
```
Comment (kudinov-fedor) [inline tests/irepela/homework_10/test_context_manager_try_except.py:12]: check negative case where you do not catch error

**Review by kudinov-fedor — DISMISSED** [review]

File: irepela/homework_10/generators.py (hunk 1 above)
Comment (winner6) [reply irepela/homework_10/generators.py]: Fixed

File: irepela/homework_10/context_manager_timer.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+    def __init__(self):
+        self.total = 0
+
+    def __enter__(self):
+        self.start = time.time()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.total += (time.time() - self.start)
```
[the remark this answers is not in the document]
Comment (winner6) [reply irepela/homework_10/context_manager_timer.py:15]: Added reset on enter

File: tests/irepela/homework_10/test_context_manager_timer.py (hunk 7 above)
Comment (winner6) [reply tests/irepela/homework_10/test_context_manager_timer.py:16]: I'm now resetting it on enter - so should be as is assert timer.total >= 2

File: irepela/homework_10/iterators.py (hunk 5 above)
Comment (winner6) [reply irepela/homework_10/iterators.py]: Removed while loop

File: irepela/homework_10/iterators.py (hunk 6 above)
Comment (winner6) [reply irepela/homework_10/iterators.py]: Removed while loop

File: tests/irepela/homework_10/test_context_manager_try_except.py (hunk 8 above)
Comment (winner6) [reply tests/irepela/homework_10/test_context_manager_try_except.py:12]: Added case without exception

**Review by kudinov-fedor — APPROVED** [review]

**Pull request merged**